src/v/config/tests/retention_property_test.cc

```diff
@@ -0,0 +1,59 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "config/config_store.h"
+#include "config/property.h"
+
+#include <seastar/testing/thread_test_case.hh>
+
+#include <chrono>
+
+namespace {
+using namespace std::chrono_literals;
+
+struct test_config : public config::config_store {
+    config::retention_property valid_retention;
+    config::retention_property inf_retention;
+
+    test_config()
+      : valid_retention(
+        *this,
+        "valid_retention",
+        "A positive value",
+        {.needs_restart = config::needs_restart::no,
+         .visibility = config::visibility::user},
+        10080min)
+      , inf_retention(
+          *this,
+          "inf_retention",
+          "-1 should be interpreted as infinity",
+          {.needs_restart = config::needs_restart::no,
+           .visibility = config::visibility::user},
+          10080min) {}
+};
+
+SEASTAR_THREAD_TEST_CASE(test_retention_property) {
+    auto cfg = test_config();
+
+    cfg.valid_retention.set_value(YAML::Load("4095"));
+    BOOST_CHECK(cfg.valid_retention() == std::chrono::milliseconds(4095));
+
+    cfg.valid_retention.set_value(std::chrono::milliseconds(809098));
+    BOOST_CHECK(cfg.valid_retention() == std::chrono::milliseconds(809098));
+
+    cfg.inf_retention.set_value(YAML::Load("-1"));
+    BOOST_CHECK(
+      cfg.inf_retention() == std::chrono::milliseconds(9223372036854775807));
+
+    cfg.inf_retention.set_value(std::chrono::milliseconds(-1));
```

Comment:
TIL that time can be negative.

Reply:
I'm not a fan of representing `inf` like this, but it's what we have, and is not uncommon in the Kafka protocol.